bittide/src/Bittide/Transceiver/WordAlign.hs

```diff
@@ -0,0 +1,178 @@
+-- SPDX-FileCopyrightText: 2024 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE NamedFieldPuns #-}
+
+-- | Fundamentally, our transceivers are sending over single bits. Through transceiver
+-- IP we can send words (e.g., 32 bits) on one end and receive words on the other
+-- end. The IP makes sure that any received are byte aligned. That means that a
+-- stream:
+--
+-- > ----------------------------------
+-- > | B3 B2 B1 B0 | G3 G2 G1 G0 | ...
+-- > ----------------------------------
+--
+-- ..might be received on the other end as:
+--
+-- > -----------------------------------------------
+-- > | .. .. B3 B2 | B1 B0 B3 B2 | B1 B0 B3 B2 | ...
+-- > -----------------------------------------------
+
+-- > -----------------------------------------------
+-- > | .. .. B0 B1 | B2 B0 B3 B2 | B1 B0 B3 B2 | ...
+-- > -----------------------------------------------
+--
+-- or by any other shift (or none at all!). This module provides utilities to end
+-- up with a word aligned stream. The basic idea is that, while \"booting\" the
+-- connection the MSB of each byte is reserved, where an asserted MSB indicates
+-- the start of a word. The other bits can be used to detect a valid connection
+-- through PRBS streams.
+--
+-- TODO: Remove this module in favor of retry logic. That is, we can keep
+--       resetting the transceivers until they're aligned. Its unclear how exactly
+--       this should work though.
+module Bittide.Transceiver.WordAlign
+  ( alignBytesFromMsbs
+  , align
+
+  -- * Convenience functions
+  , alignSymbol
+  , splitMsbs
+  , joinMsbs
+  , dealignBytes
+  ) where
+
+import Clash.Prelude
+
+import Bittide.SharedTypes (Bytes)
+import Clash.Explicit.Prelude (noReset)
+import Data.Bifunctor (Bifunctor (bimap))
+import Data.Maybe (fromMaybe)
+import Data.Tuple.Extra (curry3)
+
+data State n m = State { prev :: BitVector (n * m), offset :: Index n }
+  deriving (Show, Generic, ShowX, NFDataX)
+
+defState :: (KnownNat n, KnownNat m) => State n m
+defState = State{prev=0, offset=0}
+
+alignSymbol :: forall n . (KnownNat n, 1 <= n) => BitVector n
+alignSymbol = bit (natToNum @(n - 1))
```

Comment:
`1 +>>. 0` may be a bit more straightforward to read (and you can remove the explicit forall).

Reply:
Ah clever, I like it.